src/cluster.c

```diff
@@ -2987,7 +2979,6 @@ void clusterSendPing(clusterLink *link, int type) {
             if (node->flags & CLUSTER_NODE_NOADDR) continue;
             if (!(node->flags & CLUSTER_NODE_PFAIL)) continue;
             clusterSetGossipEntry(hdr,gossipcount,node);
-            freshnodes--;
```

Comment:
why did you remove this? just because it's not used later?

Reply:
yes, it is no longer used later

Reply:
but had someone some day use it later in that function, is this decrement correct? i think we should keep it.

Reply:
You're right, just removing it might throw an error in the future. But it's also weird to keep this code. We may re-examine the meaning of the freshnodes variable. Please refer the history commit: [commit](https://github.com/redis/redis/commit/02777bb2529a403b50be4621a93fd65cdce14343). If we always add fail nodes, I think the number of pfail nodes should be excluded from freshnode count, so that the code will be easier to understand.  it also may save the cpu loops in some state. 

```c
void clusterSendPing(clusterLink *link, int type) {
    int freshnodes = dictSize(server.cluster->nodes)-2;
...
    while(freshnodes > 0 && gossipcount < wanted && maxiterations--) {
        ...
        if (this->flags & CLUSTER_NODE_PFAIL) continue; <----when cluster have fail node, freshnodes will never be zero
        ...
    }
...
}

Reply:
@madolson Sorry to bother you, if you have time, can you help me review this pr

Reply:
Freshnodes is not "correct" anyways, it's just an upper bound approximations for the maximum number of nodes we might gossip about. I'm not sure what you would use it for. It's used primarily to prevent clusters which are all in "handshake" state from using too much time. Although, the max iterations still exists to dampen that, so it might not practically be useful anyways.